API/CopilotService/__init__.py

```diff
@@ -77,31 +79,70 @@ async def _retrieve_context_from_learn_knowledge_index(question):
     system_msg = os.environ.get("OPENAI_SPLIT_TASK_MSG", default=DEFAULT_SPLIT_TASK_MSG)
     generate_results = gpt_generate(system_msg, question, history_msg=[])
     try:
-        task_list = _build_scenario_response(generate_results)
+        raw_task_list = _build_scenario_response(generate_results)
     except Exception as e:
         logger.error(f"Error while parsing the generate results: {generate_results}, {e}")
+        return None, None
 
-    splited_tasks = []
-    for task_info in task_list:
-        # use task command as task info when length of task info > 1, otherwise use task desc as task info
-        task_info = task_info.split("||")
-        task_info = task_info[1] if len(task_info) > 1 else task_info[0]
-        splited_tasks.append(asyncio.create_task(retrieve_chunks_for_atomic_task(task_info)))
+    context_tasks = [asyncio.create_task(_build_task_context(raw_task)) for raw_task in raw_task_list]
 
-    chunks_list = []
-    if len(splited_tasks) > 0:
-        chunks_list = await asyncio.gather(*splited_tasks)
+    context_info_list = await asyncio.gather(*context_tasks)
+    task_list = [context_info[0] for context_info in context_info_list]
+    chunk_list = _join_chunks_in_context(context_info_list)
 
+    chunk_list = merge_chunks_by_command(chunk_list)
     # TODO The logic of filtering, ranking, and aggregating chunks
 
-    return task_list, chunks_list
-
-
-def _add_context_to_queston(question, task_list, usage_context):
+    return task_list, chunk_list
+
+
+async def _build_task_context(raw_task):
+    """
+    Build a guide step and context command from a raw task
+    Args:
+        raw_task: the raw task generated by GPT(prompt for splitting task)
+    Returns: A tuple. The first element is the guide step in str and
+    the second element is a list of chunks that contain some possible commands to use.
+    """
+    desc = raw_task.split("||")[0]
+    if len(raw_task.split("||")) > 1:
+        # If GPT provides a command when splitting task, validate the command.
+        cmd = raw_task.split("||")[1]
+        validate_result = validate_command_in_task(cmd)
+        if validate_result is None:
+            # If the GPT-provided command is valid, use it as a guide step.
+            task = cmd
+            chunks = []
+        elif 'Unknown Command' in validate_result.msg or 'not an Azure CLI command' in validate_result.msg:
+            # If the GPT-provided command has an error in command signature,
+            # retrieve context chunks according to the description
+            task = desc
+            chunks = await retrieve_chunks_for_atomic_task(desc)
```

Comment:
In fact, due to the excessive flexibility and diversity of natural language descriptions, using the command signature with hallucination issues to retrieve may have a more accurate effect than using descriptions, because hallucination command signatures preserve the semantics of operations without attention issues, and there are only subtle differences between the hallucination and correct command signatures.

```suggestion
            chunks = await retrieve_chunks_for_atomic_task(cmd)
```

Reply:
There are two conditions in this case. Maybe the two cases should be handled separately. 

- If the generated command is an unknown command, which means there is a hallucination issue, we should use the command signature to retrieve chunks. 
- If the generated command is not a CLI command like `git clone`, `kubecli xxx`, we should use description to retrieve chunks.

Do you think this makes sense?

Reply:
Yes, make sense, good point!